src/ts_generator.py

```diff
@@ -0,0 +1,364 @@
+        self.LOG = [0] * self.log_size
+        # same but only for PO; necessary to ensure maximum and minimum PO is respected
+        self.LOGP = [0] * self.log_size
+
+        # pogramed and forced outage rate
+        self.lf = np.empty(days_per_year, dtype=float)
+        self.lp = np.empty(days_per_year, dtype=float)
+
+        ## ???
+        self.ff = np.empty(days_per_year, dtype=float)  # ff = lf / (1 - lf)
+        self.pp = np.empty(days_per_year, dtype=float)  # pp = lp / (1 - lp)
+
+        # precalculated value to speed up generation of random outage duration
+        self.af = np.empty(days_per_year, dtype=float)
+        self.bf = np.empty(days_per_year, dtype=float)
+        self.ap = np.empty(days_per_year, dtype=float)
+        self.bp = np.empty(days_per_year, dtype=float)
+
+    def prepare_outage_duration_constant(
+        self,
+        law: ProbilityLaw,
+        volatility: float,
+        A: np.ndarray[Any, np.dtype[Any]],
+        B: np.ndarray[Any, np.dtype[Any]],
+        expecs: List[int],
+    ) -> None:
+        """
+        precalculation of constant values use in generation of outage duration
+        results are stored in A and B
+        """
+        if law == ProbilityLaw.UNIFORM:
+            for day in range(self.days_per_year):
+                D = expecs[day]
+                xtemp = volatility * (D - 1)
+                A[day] = D - xtemp
+                B[day] = 2 * xtemp + 1
+        elif law == ProbilityLaw.GEOMETRIC:
+            for day in range(self.days_per_year):
+                D = expecs[day]
+                xtemp = volatility * volatility * D * (D - 1)
+                if xtemp != 0:
+                    ytemp = (sqrt(4 * xtemp + 1) - 1) / (2 * xtemp)
+                    A[day] = D - 1 / ytemp
+                    B[day] = 1 / log(1 - ytemp)
+                else:
+                    A[day] = D - 1
+                    B[day] = 0
+
+    def duration_generator(
+        self, law: ProbilityLaw, volatility: float, a: float, b: float, expec: int
+    ) -> int:
+        """
+        generation of random outage duration
+        """
+        rnd_nb = rndgenerator.next()
+        if law == ProbilityLaw.UNIFORM:
+            return int(a + rnd_nb * b)
+        elif law == ProbilityLaw.GEOMETRIC:
+            return min(int(1 + a + b * log(rnd_nb)), int(self.log_size / 2 - 1))
+
+    def generate_time_series(
+        self,
+        cluster: ThermalCluster,
+        number_of_timeseries: int,
+        output_series: List[List[float]],
+        output_outages: Optional[List[List[Tuple[int, int, int, int, int]]]] = None,
+    ) -> None:
+        """
+        generation of multiple timeseries for a given thermal cluster
+        output_series stores available power at a given time
+        output_outages stores infomation about outages begining at a given time -> (PPO, PFO, MXO, POD, FOD)
```

Comment:
We could probably use a diffrent structure for this output to make it more understandable:
for example we could use a list of pandas.Dataframe, one dataframe for each TS index. The dataframes would have columns named PPO, PFO etc, and have one line per day.